src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.Api.async.cs

```diff
@@ -0,0 +1,125 @@
+
+                    int deserializeBufferSize = bytesRemaining + bytesRead;
+                    bool isFinalBlock = (bytesRead == 0);
+                    if (ReadData(
+                        ref readerState,
+                        returnType,
+                        isFinalBlock,
+                        buffer,
+                        deserializeBufferSize,
+                        options,
+                        ref current,
+                        ref previous,
+                        ref arrayIndex))
+                    {
+                        return (T)current.ReturnValue;
+                    }
+
+                    // We have to shift or expand the buffer because there wasn't enough data to complete deserialization.
+                    Debug.Assert(isFinalBlock == false);
+                    int bytesConsumed = (int)readerState.BytesConsumed;
+                    bytesRemaining = deserializeBufferSize - bytesConsumed;
+
+                    if (bytesRemaining <= (bufferSize / 2))
+                    {
+                        // We have less than half the buffer available, double the buffer size.
+                        bufferSize = (bufferSize < HalfMaxValue) ? bufferSize * 2 : int.MaxValue;
+
+                        byte[] dest = ArrayPool<byte>.Shared.Rent(bufferSize);
+                        bufferSize = dest.Length;
+                        if (bytesRemaining > 0)
+                        {
+                            // Copy the unprocessed data to the new buffer while shifting the processed bytes.
+                            Buffer.BlockCopy(buffer, bytesConsumed, dest, 0, bytesRemaining);
+                        }
+                        ArrayPool<byte>.Shared.Return(buffer, clearArray:true);
+                        buffer = dest;
+                    }
+                    else if (bytesRemaining > 0)
+                    {
+                        // Shift the processed bytes to the beginning of buffer to make more room.
+                        Buffer.BlockCopy(buffer, bytesConsumed, buffer, 0, bytesRemaining);
+                    }
+                } while (bytesRead > 0);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
+            }
+
+            throw new InvalidOperationException("todo");
+        }
+
+        private static bool ReadData(
+            ref JsonReaderState readerState,
+            Type returnType,
+            bool isFinalBlock,
+            byte[] buffer,
+            int bytesToRead,
+            JsonConverterSettings options,
+            ref JsonObjectState current,
+            ref List<JsonObjectState> previous,
+            ref int arrayIndex)
+        {
+            Utf8JsonReader reader = new Utf8JsonReader(new ReadOnlySequence<byte>(buffer, 0, bytesToRead), isFinalBlock, readerState);
```

Comment:
I wouldn't create a ReadOnlySequence frm the input buffer. Instead pass it in as a span slice.

```C#
var reader = new Utf8JsonReader(buffer.AsSpan(0, bytesToRead), isFinalBlock, readerState);
```